File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/session/SessionControllerImpl.java
package com.snowplowanalytics.snowplow.internal.session;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;
import androidx.core.util.Consumer;

import com.snowplowanalytics.snowplow.controller.SessionController;
import com.snowplowanalytics.snowplow.internal.Controller;
import com.snowplowanalytics.snowplow.internal.tracker.ServiceProviderInterface;
import com.snowplowanalytics.snowplow.internal.tracker.Tracker;
import com.snowplowanalytics.snowplow.internal.tracker.Logger;
import com.snowplowanalytics.snowplow.tracker.SessionState;
import com.snowplowanalytics.snowplow.util.TimeMeasure;

import java.util.concurrent.TimeUnit;

@RestrictTo(RestrictTo.Scope.LIBRARY)
public class SessionControllerImpl extends Controller implements SessionController {
    private final String TAG = SessionControllerImpl.class.getName();

    // Constructors

    public SessionControllerImpl(@NonNull ServiceProviderInterface serviceProvider) {
        super(serviceProvider);
    }

    // Control methods

    @Override
    public void pause() {
        getDirtyConfig().isPaused = true;
        getTracker().pauseSessionChecking();
    }

    @Override
    public void resume() {
        getDirtyConfig().isPaused = false;
        getTracker().resumeSessionChecking();
    }

    @Override
    public void startNewSession() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return;
        }
        session.startNewSession();
    }

    // Getters and Setters

    @Override
    public int getSessionIndex() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return -1;
        }
        return session.getSessionIndex();
    }

    @NonNull
    @Override
    public String getSessionId() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return "";
        }
        return session.getState().getSessionId();
    }

    @NonNull
    @Override
    public String getUserId() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return "";
        }
        return session.getUserId();
    }

    @Override
    public boolean isInBackground() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return false;
        }
        return session.isBackground();
    }

    @Override
    public int getBackgroundIndex() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return -1;
        }
        return session.getBackgroundIndex();
    }

    @Override
    public int getForegroundIndex() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return -1;
        }
        return session.getForegroundIndex();
    }

    @NonNull
    @Override
    public TimeMeasure getForegroundTimeout() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return new TimeMeasure(0, TimeUnit.SECONDS);
        }
        return new TimeMeasure(session.getForegroundTimeout(), TimeUnit.MILLISECONDS);
    }

    @Override
    public void setForegroundTimeout(@NonNull TimeMeasure foregroundTimeout) {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return;
        }
        getDirtyConfig().foregroundTimeout = foregroundTimeout;
        getDirtyConfig().foregroundTimeoutUpdated = true;
        session.setForegroundTimeout(foregroundTimeout.convert(TimeUnit.MILLISECONDS));
    }

    @NonNull
    @Override
    public TimeMeasure getBackgroundTimeout() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return new TimeMeasure(0, TimeUnit.SECONDS);
        }
        return new TimeMeasure(session.getBackgroundTimeout(), TimeUnit.MILLISECONDS);
    }

    @Override
    public void setBackgroundTimeout(@NonNull TimeMeasure backgroundTimeout) {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return;
        }
        getDirtyConfig().backgroundTimeout = backgroundTimeout;
        getDirtyConfig().backgroundTimeoutUpdated = true;
        session.setBackgroundTimeout(backgroundTimeout.convert(TimeUnit.MILLISECONDS));
    }

    /**
     * The callback called everytime the session is updated.
     */
    @Nullable
    @Override
    public Consumer<SessionState> getOnSessionUpdate() {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return null;
        }
        return session.onSessionUpdate;
    }

    /**
     * The callback called everytime the session is updated.
     */
    @Override
    public void setOnSessionUpdate(@Nullable Consumer<SessionState> onSessionUpdate) {
        Session session = getSession();
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled");
            return;
        }
        session.onSessionUpdate = onSessionUpdate;
    }

    // Service method

    public boolean isEnabled() {
        return getTracker().getSession() != null;
    }

    // Private methods

    @NonNull
    private Tracker getTracker() {
        return serviceProvider.getOrMakeTracker();
    }

    @Nullable
    private Session getSession() {
        return serviceProvider.getOrMakeTracker().getSession();
    }

    @NonNull
    private SessionConfigurationUpdate getDirtyConfig() {
        return serviceProvider.getSessionConfigurationUpdate();
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/session/SessionControllerImpl.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.core.session

import androidx.annotation.RestrictTo
import androidx.core.util.Consumer
import com.snowplowanalytics.core.Controller
import com.snowplowanalytics.core.tracker.Logger
import com.snowplowanalytics.core.tracker.ServiceProviderInterface
import com.snowplowanalytics.core.tracker.Tracker
import com.snowplowanalytics.snowplow.configuration.SessionConfiguration
import com.snowplowanalytics.snowplow.controller.SessionController
import com.snowplowanalytics.snowplow.tracker.SessionState
import com.snowplowanalytics.snowplow.util.TimeMeasure
import java.util.concurrent.TimeUnit

@RestrictTo(RestrictTo.Scope.LIBRARY)
class SessionControllerImpl  // Constructors
    (serviceProvider: ServiceProviderInterface) : Controller(serviceProvider), SessionController {
    private val TAG = SessionControllerImpl::class.java.name

    // Control methods
    override fun pause() {
        dirtyConfig.isPaused = true
        tracker.pauseSessionChecking()
    }

    override fun resume() {
        dirtyConfig.isPaused = false
        tracker.resumeSessionChecking()
    }

    override fun startNewSession() {
        val session = session
        if (session == null) {
            Logger.track(TAG, "Attempt to access SessionController fields when disabled")
            return
        }
        session.startNewSession()
    }

    
    // Getters and Setters
    override val sessionIndex: Int?
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return -1
            }
            return session.sessionIndex
        }
    
    override val sessionId: String?
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return ""
            }
            return session.state?.sessionId
        }
    
    override val userId: String
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return ""
            }
            return session.userId
        }
    
    override val isInBackground: Boolean
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return false
            }
            return session.isBackground
        }
    
    override val backgroundIndex: Int
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return -1
            }
            return session.backgroundIndex
        }
    
    override val foregroundIndex: Int
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return -1
            }
            return session.foregroundIndex
        }

    override var foregroundTimeout: TimeMeasure
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return TimeMeasure(0, TimeUnit.SECONDS)
            }
            return TimeMeasure(session.foregroundTimeout, TimeUnit.MILLISECONDS)
        }
        set(foregroundTimeout) {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return
            }
            dirtyConfig.foregroundTimeout = foregroundTimeout
            session.foregroundTimeout = foregroundTimeout.convert(TimeUnit.MILLISECONDS)
        }
    
    override var backgroundTimeout: TimeMeasure
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return TimeMeasure(0, TimeUnit.SECONDS)
            }
            return TimeMeasure(session.backgroundTimeout, TimeUnit.MILLISECONDS)
        }
        set(backgroundTimeout) {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return
            }
            dirtyConfig.backgroundTimeout = backgroundTimeout
            session.backgroundTimeout =
                backgroundTimeout.convert(TimeUnit.MILLISECONDS)
        }
    
    override var onSessionUpdate: Consumer<SessionState>?
        get() {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return null
            }
            return session.onSessionUpdate
        }
        set(onSessionUpdate) {
            val session = session
            if (session == null) {
                Logger.track(TAG, "Attempt to access SessionController fields when disabled")
                return
            }
            session.onSessionUpdate = onSessionUpdate
        }

    // Service method
    val isEnabled: Boolean
        get() = tracker.session != null

    // Private methods
    private val tracker: Tracker
        get() = serviceProvider.getOrMakeTracker()
    private val session: Session?
        get() = serviceProvider.getOrMakeTracker().session
    private val dirtyConfig: SessionConfiguration
        get() = serviceProvider.sessionConfiguration
}
